internal/wrphandlers/qos/qos.go

```diff
@@ -0,0 +1,138 @@
+// SPDX-FileCopyrightText: 2024 Comcast Cable Communications Management, LLC
+// SPDX-License-Identifier: Apache-2.0
+
+package qos
+
+import (
+	"context"
+	"errors"
+
+	"github.com/xmidt-org/wrp-go/v3"
+	"github.com/xmidt-org/xmidt-agent/internal/websocket"
+	"github.com/xmidt-org/xmidt-agent/internal/wrpkit"
+)
+
+var (
+	ErrInvalidInput     = errors.New("invalid input")
+	ErrMisconfiguredQOS = errors.New("misconfigured QOS")
+)
+
+// Option is a functional option type for QOS.
+type Option interface {
+	apply(*Handler) error
+}
+
+type optionFunc func(*Handler) error
+
+func (f optionFunc) apply(c *Handler) error {
+	return f(c)
+}
+
+type serviceQOSHandler func(wrp.Message) (<-chan wrp.Message, <-chan struct{})
+
+// Handler queues incoming messages and sends them to the next wrphandler
+type Handler struct {
+	// queue for wrp messages, ingested by serviceQOS
+	queue chan wrp.Message
+	// maxHeapSize is the allowable max size of the qos' priority queue, based on the sum of all queued wrp message's payload
+	maxHeapSize int
+}
+
+// New creates a new instance of the Handler struct.  The parameter next is the
+// handler that will be called and monitored for errors.
+func New(next websocket.Egress, opts ...Option) (*Handler, func(), error) {
+	if next == nil {
+		return nil, nil, ErrInvalidInput
+	}
+
+	q := make(chan wrp.Message)
+	h := &Handler{
+		queue: q,
+	}
+
+	for _, opt := range opts {
+		if opt != nil {
+			if err := opt.apply(h); err != nil {
+				return nil, nil, err
+			}
+		}
+	}
+
+	// shutdown() is used to stop serviceQOS by closing its `done` chan.
+	ctx, shutdown := context.WithCancel(context.Background())
+	go serviceQOS(ctx.Done(), h.queue, h.maxHeapSize, handleWRPWrapper(next))
```

Comment:
Starting a goroutine here may cause issues with `fx`.  This may cause the constructor `New` to take longer than `fx` wants to wait, especially in a constrained environment.  It might be better to bind the lifecycle of a `Handler` to the enclosing `fx.App`.

Reply:
that makes sense, I'll go ahead and re-introduce the `Start` and `End` methods and add some atomic state management around starting/stopping the qos goroutine